programs/cross-pile/src/lib.rs

```diff
@@ -1,319 +1,767 @@
+    #[account(
+        mut,
+        constraint = initiator_tokens_vault.key() == challenge.initiator_tokens_vault,
+    )]
+    initiator_tokens_vault: Box<Account<'info, TokenAccount>>,
+    // account to receive initiator's own bet back into
+    #[account(
+        mut,
+        constraint = initiator_own_tokens_taker.owner == challenge.initiator,
+    )]
+    initiator_own_tokens_taker: Box<Account<'info, TokenAccount>>,
 
-        let signer = &[
-            &coin_seeds[..]
-        ];
+    // application level accounts
+    pub token_program: Program<'info, Token>,
+}
 
-        let cpi_context = CpiContext::new_with_signer(
-            ctx.accounts.solrand_program.clone(),
-            cpi_accounts,
-            signer
-        );
+#[derive(Accounts)]
+pub struct CancelAfterAcceptor<'info> {
+    #[account(
+        mut,
+        constraint = initiator.key() == challenge.initiator @ ErrorCode::Unauthorized
+    )]
+    pub initiator: Signer<'info>,
+    #[account(mut, close = initiator)]
+    pub challenge: Account<'info, Challenge>,
+    #[account(
+        mut,
+        constraint = initiator_tokens_vault.key() == challenge.initiator_tokens_vault,
+    )]
+    initiator_tokens_vault: Box<Account<'info, TokenAccount>>,
+    // account to receive initiator's own bet back into
+    #[account(
+        mut,
+        constraint = initiator_own_tokens_taker.owner == challenge.initiator,
+    )]
+    initiator_own_tokens_taker: Box<Account<'info, TokenAccount>>,
 
-        solrand::cpi::transfer_authority(cpi_context)?;
+    #[account(
+        mut,
+        constraint = challenge.acceptor == acceptor.key()
+    )]
+    /// CHECK: constraint makes sure that the acceptor of the challenge is the account passed in here
+    acceptor: AccountInfo<'info>,
+    #[account(
+        mut,
+        constraint = acceptor_tokens_vault.key() == challenge.acceptor_tokens_vault,
+    )]
+    acceptor_tokens_vault: Box<Account<'info, TokenAccount>>,
+    // account to receive initiator's own bet back into
+    #[account(
+        mut,
+        constraint = acceptor_own_tokens_taker.owner == challenge.acceptor,
+    )]
+    acceptor_own_tokens_taker: Box<Account<'info, TokenAccount>>,
 
-        return Ok(());
-    }
+    // application level accounts
+    pub token_program: Program<'info, Token>,
 }
 
+// PDA that holds the state of the challenge
+#[account]
+pub struct Challenge {
+    pub initiator: Pubkey,
+    pub initiator_tokens_mint: Pubkey,
+    pub initiator_tokens_vault: Pubkey,
+    pub initiator_wager_token_amount: u64,
+    pub acceptor: Pubkey,
+    pub acceptor_tokens_mint: Pubkey,
+    pub acceptor_tokens_vault: Pubkey,
+    pub acceptor_wager_token_amount: u64,
+    pub acceptor_wager_approved: bool,
+    pub requester: Pubkey,
+    pub bump: u8,
+}
+
+// arguments list for new_challenge
+// TODO: make the challenge bump include a UUID to allow initiators to create as many challenges as they want
```

Comment:
Can you add the `requester` key here to the `seeds` ie: `seeds = [b"coin-seed".as_ref(), initiator.key().as_ref(), requester.key().as_ref()]` . Since the requester takes a uuid, this will effectively enable a user to create as many challenges as they want.

Reply:
Yes, I'll add the requester key 👍